### PR TITLE
[Validator][RecursiveContextualValidator] Prevent validated hash collisions

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | -
| Deprecations? | -
| Tickets       | -
| License       | MIT
| Doc PR        | -

`RecursiveContextualValidator` uses `spl_object_hash` at different places to generate cache keys. The problem with this method is that there can be collisions (same hash for two different objects) since once an object is destroyed, `spl_object_hash` can return a previously used hash.

I encountered the problem on a project with constraints that are created on the fly. The first constraint is applied and is "cached". Then it is destroyed. Then another constraint happen to have the exact same `spl_object_hash`. It is skipped while it would cause violations. I was able to reproduce the problem consistently in the unit test.

The easy way to deal with this is to keep references to all objects (proposed solution). But I guess it could leak too much.

We could also try to "clean" the `ExecutionContext` stores at the end of the three validate methods. However, we would still need to keep references to the initialized objects since they must really be initialized one time, even on multiple calls

### CODE REVIEW

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php [hunk 1]
```diff
@@ -457,7 +465,15 @@ private function validateClassNode($object, $cacheKey, ClassMetadataInterface $m
             $defaultOverridden = false;
 
             // Use the object hash for group sequences
-            $groupHash = \is_object($group) ? spl_object_hash($group) : $group;
+            if (\is_object($group)) {
+                $groupHash = spl_object_hash($group);
+
+                if (!isset($this->validatedObjectsReferences[$groupHash])) {
```
Comment (HeahDude) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:471]: Why is this needed for group sequences?

**Comment by nicolas-grekas** [comment]
We should find a place where these arrays can be emptied. The memory leak should be avoided.

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php (hunk 1 above)
Comment (fancyweb) [reply src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:471]: GroupSequences can suffer from the same problem since they can be passed directly to the validate method as objects.

**Comment by fancyweb** [comment]
> We should find a place where these arrays can be emptied. The memory leak should be avoided.

I think we can clear constraints and group sequences arrays at the end of each validate methods. However, we can never clean the initialized objects. Would that work for you?

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php [hunk 2]
```diff
@@ -457,7 +475,15 @@ private function validateClassNode($object, $cacheKey, ClassMetadataInterface $m
             $defaultOverridden = false;
 
             // Use the object hash for group sequences
-            $groupHash = \is_object($group) ? spl_object_hash($group) : $group;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:460]: would replacing this line by `$groupHash = serialize($group);` fix the issue?

**Comment by fabpot** [comment]
@fancyweb What's the status of this PR?

**Comment by fancyweb** [comment]
I need to check Nicolas's suggestion, please keep it open.

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php (hunk 2 above)
Comment (fancyweb) [reply src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:461]: It works as long as one does not add constraints to class metadata between 2 validations on the same group. I guess it's hacky enough to be ignored.

**Comment by fancyweb** [comment]
> We should find a place where these arrays can be emptied. The memory leak should be avoided.

> I think we can clear constraints and group sequences arrays at the end of each validate methods. However, we can never clean the initialized objects. Would that work for you?

Unfortunately that does not work. I had to remove this part.

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php (hunk 2 above)
Comment (fancyweb) [reply src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:461]: I added a test for that btw.

**Comment by fancyweb** [comment]
@xabbuh I looked at this a few days ago but it needs more work. My patch does not fix all cases because it does not take into account the current defaultPropertyPath. One of FormValidator tests fails because of this. You can have a look if you want to.

**Comment by xabbuh** [comment]
continued in #38387, thanks for starting the work on this @fancyweb
